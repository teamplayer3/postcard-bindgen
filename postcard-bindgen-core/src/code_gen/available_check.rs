use super::variable_path::{VariableAccess, VariablePath};

#[derive(Clone)]
pub enum AvailableCheck<L> {
    Object(VariablePath<L>, String),
    None,
}

impl<L> AvailableCheck<L> {
    pub fn from_variable_path(path: VariablePath<L>) -> Self {
        let (path, last) = path.modify_pop();
        match last {
            Some(VariableAccess::Field(name)) => Self::Object(path, name),
            _ => Self::None,
        }
    }
}
