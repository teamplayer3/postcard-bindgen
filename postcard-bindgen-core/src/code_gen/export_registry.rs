use genco::{lang::Lang, quote, tokens::FormatInto, Tokens};

#[derive(Debug, Default, Clone)]
pub enum ExportMode {
    #[default]
    Cjs,
    Esm,
}

pub struct ExportRegistry<L>
where
    L: Lang,
{
    pub(super) exports: Vec<Tokens<L>>,
    pub(super) export_mode: ExportMode,
}

impl<L> ExportRegistry<L>
where
    L: Lang,
{
    pub fn new(export_mode: ExportMode) -> Self {
        Self {
            exports: Vec::new(),
            export_mode,
        }
    }
}

impl<L> ExportRegistry<L>
where
    L: Lang,
{
    pub fn push(&mut self, export: impl FormatInto<L>) {
        self.exports.push(quote!($export));
    }
}
