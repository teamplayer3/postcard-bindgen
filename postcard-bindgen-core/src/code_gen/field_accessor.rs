#[derive(Debug, Clone, Copy)]
pub enum FieldAccessor<'a> {
    Object(&'a str),
    Array,
    None,
}
