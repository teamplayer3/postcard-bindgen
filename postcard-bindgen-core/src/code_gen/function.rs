use genco::{lang::Lang, quote, tokens::FormatInto, Tokens};

#[macro_export]
macro_rules! function_args {
    ($(($n:expr, $t:expr)),+ $(,)?) => {

        [$(({
            use genco::tokens::FormatInto;
            let mut tokens = genco::Tokens::new();
            $n.format_into(&mut tokens);
            tokens
        }, {
            use genco::tokens::FormatInto;
            let mut tokens = genco::Tokens::new();
            $t.format_into(&mut tokens);
            tokens
        })),+]
    };
    ($($x:expr),+ $(,)?) => {

        [$({
            use genco::tokens::FormatInto;
            let mut tokens = genco::Tokens::new();
            $x.format_into(&mut tokens);
            tokens
        }),+]
    };

}

pub struct FunctionArg<L>
where
    L: Lang,
{
    pub(super) name: Tokens<L>,
    pub(super) r#type: Option<Tokens<L>>,
}

impl<L> FunctionArg<L>
where
    L: Lang,
{
    pub fn new(name: impl FormatInto<L>, r#type: impl FormatInto<L>) -> Self {
        Self {
            name: quote!($name),
            r#type: Some(quote!($r#type)),
        }
    }

    pub fn new_untyped(name: impl FormatInto<L>) -> Self {
        Self {
            name: quote!($name),
            r#type: None,
        }
    }
}

impl<N, T, L> From<(N, T)> for FunctionArg<L>
where
    L: Lang,
    N: FormatInto<L>,
    T: FormatInto<L>,
{
    fn from((name, r#type): (N, T)) -> Self {
        Self::new(name, r#type)
    }
}

impl<L> From<Tokens<L>> for FunctionArg<L>
where
    L: Lang,
{
    fn from(name: Tokens<L>) -> Self {
        Self::new_untyped(name)
    }
}

pub trait ToArgs<L>
where
    L: Lang,
{
    fn to_args(self) -> Vec<FunctionArg<L>>;
}

impl<L, A> ToArgs<L> for Vec<A>
where
    L: Lang,
    A: Into<FunctionArg<L>>,
{
    fn to_args(self) -> Vec<FunctionArg<L>> {
        self.into_iter().map(|a| a.into()).collect()
    }
}

impl<L> ToArgs<L> for FunctionArg<L>
where
    L: Lang,
{
    fn to_args(self) -> Vec<FunctionArg<L>> {
        vec![self]
    }
}

impl<L> ToArgs<L> for ()
where
    L: Lang,
{
    fn to_args(self) -> Vec<FunctionArg<L>> {
        Vec::new()
    }
}

impl<L, F, const N: usize> ToArgs<L> for [F; N]
where
    L: Lang,
    F: Into<FunctionArg<L>>,
{
    fn to_args(self) -> Vec<FunctionArg<L>> {
        self.into_iter().map(|f| f.into()).collect()
    }
}

impl<'a, L, F, const N: usize> ToArgs<L> for &'a [F; N]
where
    L: Lang,
    &'a F: Into<FunctionArg<L>>,
{
    fn to_args(self) -> Vec<FunctionArg<L>> {
        self.iter().map(|f| f.into()).collect()
    }
}

pub struct Function<L>
where
    L: Lang,
{
    pub(super) args: Vec<FunctionArg<L>>,
    pub(super) name: Tokens<L>,
    pub(super) body: Tokens<L>,
    pub(super) return_type: Option<Tokens<L>>,
    pub(super) doc_string: Option<String>,
}

impl<L> Function<L>
where
    L: Lang,
{
    pub fn new(
        name: impl FormatInto<L>,
        args: impl ToArgs<L>,
        body: impl FormatInto<L>,
        return_type: impl FormatInto<L>,
    ) -> Self {
        Self {
            name: quote!($name),
            args: args.to_args(),
            return_type: Some(quote!($return_type)),
            body: quote!($body),
            doc_string: None,
        }
    }

    pub fn new_untyped(
        name: impl FormatInto<L>,
        args: impl ToArgs<L>,
        body: impl FormatInto<L>,
    ) -> Self {
        Self {
            name: quote!($name),
            args: args.to_args(),
            return_type: None,
            body: quote!($body),
            doc_string: None,
        }
    }

    pub fn with_doc_string(mut self, doc_string: impl ToString) -> Self {
        self.doc_string = Some(doc_string.to_string());
        self
    }
}
