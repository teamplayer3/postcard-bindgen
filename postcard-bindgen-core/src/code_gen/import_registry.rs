use std::{borrow::Cow, collections::HashMap};

use crate::path::PathBuf;

#[derive(Debug, Clone)]
pub enum ImportItem {
    All,
    Single(Cow<'static, str>),
    Aliased {
        item_name: Cow<'static, str>,
        alias: Cow<'static, str>,
    },
}

#[derive(Debug, PartialEq, Eq)]
pub struct Import {
    pub name: Cow<'static, str>,
    pub alias: Option<Cow<'static, str>>,
}

impl Import {
    fn new(name: Cow<'static, str>) -> Self {
        Self { name, alias: None }
    }

    fn aliased(name: Cow<'static, str>, alias: Cow<'static, str>) -> Self {
        Self {
            name,
            alias: Some(alias),
        }
    }
}

impl PartialOrd for Import {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Import {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.name.cmp(&other.name)
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum ImportMode {
    All,
    Single(Vec<Import>),
}

#[derive(Debug, Hash, Eq, PartialEq, Clone)]
pub enum Package {
    Extern(PathBuf<'static>),
    Intern(PathBuf<'static>),
    Relative(PathBuf<'static>),
}

impl PartialOrd for Package {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Package {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        match self {
            Self::Extern(name) => match other {
                Self::Extern(other_name) => name.cmp(other_name),
                _ => std::cmp::Ordering::Less,
            },
            Self::Intern(name) => match other {
                Self::Extern(_) => std::cmp::Ordering::Greater,
                Self::Intern(other_name) => name.cmp(other_name),
                Self::Relative(_) => std::cmp::Ordering::Less,
            },
            Self::Relative(name) => match other {
                Self::Relative(other_name) => name.cmp(other_name),
                _ => std::cmp::Ordering::Greater,
            },
        }
    }
}

pub struct ImportRegistry {
    imports: HashMap<Package, ImportMode>,
    base_path: String,
}

impl ImportRegistry {
    pub fn new(base_path: String) -> Self {
        Self {
            imports: HashMap::new(),
            base_path,
        }
    }

    pub fn push(&mut self, package: Package, item: ImportItem) {
        self.imports
            .entry(package)
            .and_modify(|e| match &item {
                ImportItem::All => *e = ImportMode::All,
                ImportItem::Single(item) => match e {
                    ImportMode::All => (),
                    ImportMode::Single(imports) => {
                        let import = Import::new(item.clone());
                        if !imports.contains(&import) {
                            imports.push(import);
                            imports.sort();
                        }
                    }
                },
                ImportItem::Aliased { item_name, alias } => match e {
                    ImportMode::All => (),
                    ImportMode::Single(imports) => {
                        let import = Import::aliased(item_name.clone(), alias.clone());
                        if !imports.contains(&import) {
                            imports.push(import);
                            imports.sort();
                        }
                    }
                },
            })
            .or_insert_with(|| match item {
                ImportItem::All => ImportMode::All,
                ImportItem::Single(item) => ImportMode::Single(vec![Import::new(item)]),
                ImportItem::Aliased { item_name, alias } => {
                    ImportMode::Single(vec![Import::aliased(item_name, alias)])
                }
            });
    }

    pub(super) fn into_items_sorted(self) -> (String, Vec<(Package, ImportMode)>) {
        let mut items = self.imports.into_iter().collect::<Vec<_>>();
        items.sort_by_key(|(package, ..)| package.to_owned());

        (self.base_path, items)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_into_items_sorted() {
        let mut registry = ImportRegistry::new("base_package".into());

        registry.push(
            Package::Relative("rel".into()),
            ImportItem::Single("test".into()),
        );
        registry.push(
            Package::Relative("rel".into()),
            ImportItem::Single("a_test".into()),
        );
        registry.push(
            Package::Extern("extern".into()),
            ImportItem::Single("test".into()),
        );

        registry.push(Package::Intern("package".into()), ImportItem::All);

        let (base_path, items) = registry.into_items_sorted();

        assert_eq!(base_path, "base_package");

        assert_eq!(
            items,
            vec![
                (
                    Package::Extern("extern".into()),
                    ImportMode::Single(vec![Import::new("test".into())])
                ),
                (Package::Intern("package".into()), ImportMode::All),
                (
                    Package::Relative("rel".into()),
                    ImportMode::Single(vec![
                        Import::new("a_test".into()),
                        Import::new("test".into())
                    ])
                ),
            ]
        );
    }
}
