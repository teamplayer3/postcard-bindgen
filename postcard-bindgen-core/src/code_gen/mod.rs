mod utils;

#[cfg(feature = "js")]
mod js;

#[cfg(feature = "python")]
mod python;

#[cfg(feature = "js")]
pub use js::{generate_js as generate, type_checking::ts::gen_ts_typings as generate_typings};

#[cfg(feature = "python")]
pub use python::{generate, type_checking::typing::generate_typings};