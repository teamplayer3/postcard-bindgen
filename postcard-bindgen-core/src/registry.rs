use core::fmt::Display;
use std::{borrow::Cow, collections::VecDeque};

use alloc::vec::Vec;
use tree_ds::prelude::{Node, NodeRemovalStrategy, Tree};

use crate::{
    path::Path,
    type_info::{GenJsBinding, ValueType},
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Container {
    pub path: Path<'static, 'static>,
    pub name: &'static str,
    pub r#type: BindingType,
}

impl Container {
    pub fn flatten_paths(&mut self) {
        self.path.flatten();

        match self.r#type {
            BindingType::Struct(ref mut ty) => ty.flatten_paths(),
            BindingType::Enum(ref mut ty) => ty.flatten_paths(),
            BindingType::TupleStruct(ref mut ty) => ty.flatten_paths(),
            _ => (),
        }
    }
}

pub struct ContainerInfo<'a> {
    pub name: Cow<'a, str>,
    pub path: Path<'a, 'a>,
}

impl From<&Container> for ContainerInfo<'_> {
    fn from(container: &Container) -> Self {
        Self {
            name: Cow::Borrowed(container.name),
            path: container.path.clone(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BindingType {
    Struct(StructType),
    TupleStruct(TupleStructType),
    UnitStruct(UnitStructType),
    Enum(EnumType),
}

#[derive(Debug, Default, Clone, PartialEq, Eq)]
// encoded into | variant index | (inner)
pub struct EnumType {
    pub variants: Vec<EnumVariant>,
}

impl EnumType {
    pub fn new() -> Self {
        Self::default()
    }

    // index is set based on order of variant registration
    pub fn register_variant(&mut self, name: &'static str) {
        self.variants.push(EnumVariant {
            index: self.variants.len(),
            name,
            inner_type: EnumVariantType::Empty,
        });
    }

    pub fn register_variant_tuple(&mut self, name: &'static str, fields: TupleFields) {
        self.variants.push(EnumVariant {
            index: self.variants.len(),
            name,
            inner_type: EnumVariantType::Tuple(fields.into_inner()),
        });
    }

    pub fn register_unnamed_struct(&mut self, name: &'static str, fields: StructFields) {
        self.variants.push(EnumVariant {
            index: self.variants.len(),
            name,
            inner_type: EnumVariantType::NewType(fields.into_inner()),
        })
    }

    fn flatten_paths(&mut self) {
        for variant in &mut self.variants {
            match &mut variant.inner_type {
                EnumVariantType::NewType(fields) => {
                    for field in fields {
                        field.v_type.flatten_paths();
                    }
                }
                EnumVariantType::Tuple(fields) => {
                    for field in fields {
                        field.flatten_paths();
                    }
                }
                _ => {}
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EnumVariant {
    pub index: usize,
    pub name: &'static str,
    pub inner_type: EnumVariantType,
}

impl AsRef<EnumVariant> for EnumVariant {
    fn as_ref(&self) -> &EnumVariant {
        self
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EnumVariantType {
    Empty,
    Tuple(Vec<ValueType>),
    // for unnamed structs create struct with custom name ( __EnumName_Struct1)
    NewType(Vec<StructField>),
}

#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct StructType {
    pub fields: Vec<StructField>,
}

impl StructType {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn register_field<T: GenJsBinding>(&mut self, name: &'static str) {
        self.fields.push(StructField {
            name,
            v_type: T::get_type(),
        })
    }

    fn flatten_paths(&mut self) {
        for field in &mut self.fields {
            field.v_type.flatten_paths();
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct TupleStructType {
    pub fields: Vec<ValueType>,
}

impl TupleStructType {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn register_field<T: GenJsBinding>(&mut self) {
        self.fields.push(T::get_type())
    }

    fn flatten_paths(&mut self) {
        for field in &mut self.fields {
            field.flatten_paths();
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct UnitStructType;

impl UnitStructType {
    pub fn new() -> Self {
        Self {}
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StructField {
    pub name: &'static str,
    pub v_type: ValueType,
}

#[derive(Debug, Default)]
pub struct StructFields(Vec<StructField>);

impl StructFields {
    pub fn register_field<T: GenJsBinding>(&mut self, name: &'static str) {
        self.0.push(StructField {
            name,
            v_type: T::get_type(),
        })
    }

    fn into_inner(self) -> Vec<StructField> {
        self.0
    }
}

#[derive(Default)]
pub struct TupleFields(Vec<ValueType>);

impl TupleFields {
    pub fn register_field<T: GenJsBinding>(&mut self) {
        self.0.push(T::get_type())
    }

    fn into_inner(self) -> Vec<ValueType> {
        self.0
    }
}

pub struct ContainerCollection(Tree<NodeId, NodeType>);

impl ContainerCollection {
    pub fn flatten(&mut self) {
        let root_node_id = self.0.get_root_node().unwrap().get_node_id();
        let root_node = self.0.get_node_by_id(&root_node_id).unwrap();

        let root_mod_nodes = root_node
            .get_children_ids()
            .iter()
            .filter_map(|i| {
                let node = self.0.get_node_by_id(i).unwrap();

                let node_value = node.get_value().unwrap();
                if node_value.is_module() {
                    Some(node)
                } else {
                    None
                }
            })
            .collect::<VecDeque<_>>();

        let mut mod_nodes = root_mod_nodes.clone();

        loop {
            if mod_nodes.is_empty() {
                break;
            }

            let node = mod_nodes.pop_front().unwrap();

            let children = node.get_children_ids();

            for child in children {
                let child_node = self.0.get_node_by_id(&child).unwrap();
                let child_value = child_node.get_value().unwrap();

                node.remove_child(child_node.clone());

                if child_value.is_container() {
                    root_node.add_child(child_node);
                } else if child_value.is_module() {
                    mod_nodes.push_back(child_node);
                }
            }
        }

        for node in root_mod_nodes {
            self.0
                .remove_node(
                    &node.get_node_id(),
                    NodeRemovalStrategy::RemoveNodeAndChildren,
                )
                .unwrap();
        }

        for node in root_node
            .get_children_ids()
            .iter()
            .map(|node| self.0.get_node_by_id(node).unwrap())
        {
            node.update_value(|v| v.as_mut().unwrap().container_mut().unwrap().flatten_paths());
        }
    }

    pub fn all_containers(&self) -> impl Iterator<Item = Container> + Clone + '_ {
        self.0
            .get_nodes()
            .iter()
            .filter_map(|node| node.get_value().unwrap().container().cloned())
    }

    pub fn containers_per_module(&self) -> (Vec<Container>, Vec<Module<'_>>) {
        let root_node = self.0.get_root_node().unwrap().get_node_id();
        container_and_modules_per_mod(&self.0, &root_node)
    }
}

#[derive(Debug, Clone)]
pub struct Module<'a> {
    tree: &'a Tree<NodeId, NodeType>,
    node_id: NodeId,
    name: Cow<'static, str>,
    cached_path: Option<String>,
}

impl<'a> Module<'a> {
    fn new(tree: &'a Tree<NodeId, NodeType>, node_id: NodeId, name: Cow<'static, str>) -> Self {
        Self {
            tree,
            node_id,
            name,
            cached_path: None,
        }
    }

    pub fn path(&self) -> String {
        if let Some(path) = &self.cached_path {
            path.clone()
        } else {
            let mut curr_node = self.tree.get_node_by_id(&self.node_id).unwrap();
            let mut path = Vec::new();
            while let Some(id) = curr_node.get_parent_id() {
                let node = self.tree.get_node_by_id(&id).unwrap();
                if let NodeType::Module(name) = node.get_value().unwrap() {
                    if name != "::" {
                        path.push(name);
                        curr_node = node;
                    } else {
                        break;
                    }
                }
            }

            path.join("/")
        }
    }

    pub fn name(&self) -> &str {
        self.name.as_ref()
    }

    pub fn entries(&self) -> (Vec<Container>, Vec<Module<'a>>) {
        container_and_modules_per_mod(self.tree, &self.node_id)
    }
}

fn container_and_modules_per_mod<'a>(
    tree: &'a Tree<NodeId, NodeType>,
    node_id: &NodeId,
) -> (Vec<Container>, Vec<Module<'a>>) {
    let node = tree.get_node_by_id(node_id).unwrap();

    node.sort_children(|a, b| {
        let a_height = tree.get_node_height(a).unwrap();
        let b_height = tree.get_node_height(b).unwrap();

        a_height.cmp(&b_height).reverse()
    });

    let mut mods = Vec::new();
    let mut containers = Vec::new();

    for (id, child) in node
        .get_children_ids()
        .iter()
        .map(|id| (id, tree.get_node_by_id(id).unwrap()))
    {
        match child.get_value().unwrap() {
            NodeType::Module(name) => mods.push(Module::new(tree, *id, name)),
            NodeType::Container(container) => containers.push(container.clone()),
        }
    }

    (containers, mods)
}

type NodeId = u128;

enum PathExists {
    Full(NodeId),
    Partly(NodeId, Cow<'static, str>),
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum NodeType {
    Container(Container),
    Module(Cow<'static, str>),
}

impl Default for NodeType {
    fn default() -> Self {
        NodeType::Module("_".into())
    }
}

impl Display for NodeType {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            NodeType::Container(c) => write!(f, "Container({}, {})", c.name, c.path),
            NodeType::Module(m) => write!(f, "Module({})", m),
        }
    }
}

impl NodeType {
    fn container(&self) -> Option<&Container> {
        match self {
            NodeType::Container(c) => Some(c),
            _ => None,
        }
    }

    fn container_mut(&mut self) -> Option<&mut Container> {
        match self {
            NodeType::Container(c) => Some(c),
            _ => None,
        }
    }

    fn is_module(&self) -> bool {
        matches!(self, NodeType::Module(_))
    }

    fn is_container(&self) -> bool {
        matches!(self, NodeType::Container(_))
    }
}

#[derive(Debug)]
pub struct BindingsRegistry(Tree<NodeId, NodeType>);

impl BindingsRegistry {
    pub fn register_struct_binding(
        &mut self,
        name: &'static str,
        path: impl Into<Cow<'static, str>>,
        value: StructType,
    ) {
        self.insert_container(Container {
            path: Path::new(path, "::"),
            name,
            r#type: BindingType::Struct(value),
        });
    }

    pub fn register_tuple_struct_binding(
        &mut self,
        name: &'static str,
        path: impl Into<Cow<'static, str>>,
        value: TupleStructType,
    ) {
        self.insert_container(Container {
            path: Path::new(path, "::"),
            name,
            r#type: BindingType::TupleStruct(value),
        });
    }

    pub fn register_unit_struct_binding(
        &mut self,
        name: &'static str,
        path: impl Into<Cow<'static, str>>,
        value: UnitStructType,
    ) {
        self.insert_container(Container {
            path: Path::new(path, "::"),
            name,
            r#type: BindingType::UnitStruct(value),
        });
    }

    pub fn register_enum_binding(
        &mut self,
        name: &'static str,
        path: impl Into<Cow<'static, str>>,
        value: EnumType,
    ) {
        self.insert_container(Container {
            path: Path::new(path, "::"),
            name,
            r#type: BindingType::Enum(value),
        });
    }

    pub fn into_entries(self) -> ContainerCollection {
        ContainerCollection(self.0)
    }

    fn insert_container(&mut self, container: Container) {
        let mut node = self.0.get_root_node().unwrap();
        let node_id = {
            let container_path = &container.path;
            let mut parts = container_path
                .parts()
                .skip(1)
                .map(|p| p.to_owned())
                .peekable();
            let path_exists = loop {
                let part = parts.next();
                let is_last = parts.peek().is_none();

                if let Some(part) = part {
                    let node_ids = node.get_children_ids();
                    let child = node_ids.iter().find(|child| {
                        let node = self.0.get_node_by_id(child).unwrap();
                        matches!(node.get_value(), Some(NodeType::Module(p)) if p == part)
                    });

                    if let Some(child) = child {
                        if is_last {
                            break PathExists::Full(*child);
                        }

                        node = self.0.get_node_by_id(child).unwrap();
                    } else {
                        break PathExists::Partly(node.get_node_id(), part.into());
                    }
                } else {
                    break PathExists::Full(node.get_node_id());
                }
            };

            match path_exists {
                PathExists::Full(node_id) => node_id,
                PathExists::Partly(node_id, part) => {
                    let mut node = self
                        .0
                        .add_node(
                            Node::new_with_auto_id(Some(NodeType::Module(part.clone()))),
                            Some(&node_id),
                        )
                        .unwrap();

                    for part in parts {
                        node = self
                            .0
                            .add_node(
                                Node::new_with_auto_id(Some(NodeType::Module(part.clone().into()))),
                                Some(&node),
                            )
                            .unwrap();
                    }

                    node
                }
            }
        };

        self.0
            .add_node(
                Node::new_with_auto_id(Some(NodeType::Container(container))),
                Some(&node_id),
            )
            .unwrap();
    }
}

impl Default for BindingsRegistry {
    fn default() -> Self {
        let mut tree: Tree<NodeId, NodeType> = Tree::new(None);
        tree.add_node(
            Node::new_with_auto_id(Some(NodeType::Module("::".into()))),
            None,
        )
        .unwrap();
        Self(tree)
    }
}

pub trait JsBindings {
    fn create_bindings(registry: &mut BindingsRegistry);
}

#[cfg(test)]
mod test {
    use crate::registry::{
        BindingsRegistry, EnumType, JsBindings, StructFields, StructType, TupleFields,
        TupleStructType,
    };

    #[test]
    fn test_registry_struct() {
        #[allow(unused)]
        struct Test {
            a: u8,
            b: u16,
            c: &'static str,
        }

        impl JsBindings for Test {
            fn create_bindings(registry: &mut BindingsRegistry) {
                let mut ty = StructType::new();

                ty.register_field::<u8>("a");
                ty.register_field::<u16>("b");
                ty.register_field::<&str>("c");

                registry.register_struct_binding("Test", "", ty);
            }
        }

        let mut registry = BindingsRegistry::default();
        Test::create_bindings(&mut registry);
    }

    #[test]
    fn test_registry_tuple_struct() {
        #[allow(dead_code)]
        struct Test(u8, &'static str, &'static [u8]);

        impl JsBindings for Test {
            fn create_bindings(registry: &mut BindingsRegistry) {
                let mut ty = TupleStructType::new();

                ty.register_field::<u8>();
                ty.register_field::<&str>();
                ty.register_field::<&[u8]>();

                registry.register_tuple_struct_binding("Test", "", ty);
            }
        }

        let mut registry = BindingsRegistry::default();
        Test::create_bindings(&mut registry);
    }

    #[test]
    fn test_registry_enum() {
        #[allow(unused)]
        enum Test {
            A,
            B(u8),
            C { a: &'static str, b: u16 },
        }

        impl JsBindings for Test {
            fn create_bindings(registry: &mut BindingsRegistry) {
                let mut ty = EnumType::new();

                ty.register_variant("A");

                let mut fields = TupleFields::default();
                fields.register_field::<u8>();
                ty.register_variant_tuple("B", fields);

                let mut fields = StructFields::default();
                fields.register_field::<&str>("a");
                fields.register_field::<u16>("b");
                ty.register_unnamed_struct("C", fields);

                registry.register_enum_binding("Test", "", ty);
            }
        }

        let mut registry = BindingsRegistry::default();
        Test::create_bindings(&mut registry);
    }
}
