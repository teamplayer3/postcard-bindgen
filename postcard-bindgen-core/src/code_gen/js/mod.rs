use genco::{prelude::js::Tokens, quote};

use crate::registry::BindingType;

use self::{
    ser_des::{
        gen_deserialize_func, gen_ser_des_classes, gen_ser_des_functions, gen_serialize_func,
    },
    type_checking::gen_type_checkings,
};

mod generateable;
pub mod ser_des;
pub mod type_checking;
mod utils;

const JS_ENUM_VARIANT_KEY: &str = "tag";
const JS_ENUM_VARIANT_VALUE: &str = "value";
const JS_OBJECT_VARIABLE: &str = "v";

pub fn generate_js(tys: impl AsRef<[BindingType]>) -> Tokens {
    let ser_des_body = gen_ser_des_functions(&tys);
    let type_checks = gen_type_checkings(&tys);
    quote!(
        $(gen_ser_des_classes())
        $ser_des_body
        $type_checks
        $(gen_serialize_func(&tys))
        $(gen_deserialize_func(tys))
    )
}
