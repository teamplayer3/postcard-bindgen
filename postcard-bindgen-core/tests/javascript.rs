mod registry;

use postcard_bindgen_core::code_gen::js::{generate, GenerationSettings};

use registry::init_registry;

#[test]
fn test_runtime_checks() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all();
    let (exports, _meta) = generate(containers, gen_settings);

    let runtime_checks_file = exports
        .file("runtime_checks")
        .unwrap()
        .to_file_string()
        .unwrap();
    insta::assert_snapshot!(runtime_checks_file);
}

#[test]
fn test_ser_with_runtime_checks() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all();
    let (exports, _meta) = generate(containers, gen_settings);

    let ser_file = exports.file("ser").unwrap().to_file_string().unwrap();
    insta::assert_snapshot!(ser_file);
}

#[test]
fn test_ser_without_runtime_checks() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all().runtime_type_checks(false);
    let (exports, _meta) = generate(containers, gen_settings);

    let ser_file = exports.file("ser").unwrap().to_file_string().unwrap();
    insta::assert_snapshot!(ser_file);
}

#[test]
fn test_des() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all();
    let (exports, _meta) = generate(containers, gen_settings);

    let des_file = exports.file("des").unwrap().to_file_string().unwrap();
    insta::assert_snapshot!(des_file);
}

#[test]
fn test_ts_types() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all().runtime_type_checks(false);
    let (exports, _meta) = generate(containers, gen_settings);

    let ts_file = exports.file("ts").unwrap().to_file_string().unwrap();
    insta::assert_snapshot!(ts_file);
}

#[test]
fn test_ser_cjf_module() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all().esm_module(false);
    let (exports, _meta) = generate(containers, gen_settings);

    let ser_file = exports.file("ser").unwrap().to_file_string().unwrap();
    insta::assert_snapshot!(ser_file);
}

#[test]
fn test_des_cjf_module() {
    let containers = init_registry().into_entries();

    let gen_settings = GenerationSettings::enable_all().esm_module(false);
    let (exports, _meta) = generate(containers, gen_settings);

    let des_file = exports.file("des").unwrap().to_file_string().unwrap();
    insta::assert_snapshot!(des_file);
}
