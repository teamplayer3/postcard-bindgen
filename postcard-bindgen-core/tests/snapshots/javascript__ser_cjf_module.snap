---
source: postcard-bindgen-core/tests/javascript.rs
expression: ser_file
snapshot_kind: text
---
function serialize_STRUCT_TYPE(s, v) {
    s.serialize_number(U32_BYTES, false, v.field_1);
    s.serialize_string(v.field_2);
    s.serialize_array((s, v) => s.serialize_number(U32_BYTES, false, v), v.field_3);
    s.serialize_array((s, v) => serialize_STRUCT_TYPE(s, v), v.field_4);
    s.serialize_number(U32_BYTES, false, v.field_5.start);
    s.serialize_number(U32_BYTES, false, v.field_5.end);
    s.serialize_string_key_map((s, v) => s.serialize_number(U32_BYTES, false, v), v.field_6);
    s.serialize_map((d, k, v) => [
        s.serialize_number(U32_BYTES, false, k),
        s.serialize_number(U32_BYTES, false, v)
    ], v.field_7);
    if (v.field_8 !== undefined) {
        s.serialize_number(U32_BYTES, false, 1);
        s.serialize_number(U32_BYTES, false, v.field_8)
    } else {
        s.serialize_number(U32_BYTES, false, 0)
    };
    s.serialize_bool(v.field_9);
    s.serialize_number(U32_BYTES, false, v.field_10[0]);
    s.serialize_string(v.field_10[1]);
    s.serialize_array((s, v) => s.serialize_string(v), v.field_11, 3);
    s.serialize_array((s, v) => s.serialize_number(U32_BYTES, false, v), v.field_12);
    s.serialize_number_float(U32_BYTES, v.field_13);
}

function serialize_UNIT_STRUCT_TYPE(s, v) {}

function serialize_TUPLE_STRUCT_TYPE(s, v) {
    s.serialize_number(U32_BYTES, false, v[0]);
    s.serialize_string(v[1]);
}

function serialize_ENUM_TYPE(s, v) {
    switch (v.tag) {
    case "AVariant":
        s.serialize_number(U32_BYTES, false, 0);
        break;
    case "BVariant":
        s.serialize_number(U32_BYTES, false, 1);
        s.serialize_number(U32_BYTES, false, v.value[0]);
        s.serialize_string(v.value[1]);
        break;
    case "CVariant":
        s.serialize_number(U32_BYTES, false, 2);
        s.serialize_number(U32_BYTES, false, v.value.field_1);
        s.serialize_string(v.value.field_2);
        serialize_STRUCT_TYPE(s, v.value.struct_type);
        break;
    default:
        throw "variant not implemented"
    }
}

function serialize(type, value) {
    if (!(typeof type === "string")) {
        throw "type must be a string";
    }
    const s = new Serializer();
    switch (type) {
    case "StructType":
        if (is_STRUCT_TYPE(value)) {
            serialize_STRUCT_TYPE(s, value);
        } else {
            throw "value has wrong format";
        }
        break;
    case "UnitStructType":
        if (is_UNIT_STRUCT_TYPE(value)) {
            serialize_UNIT_STRUCT_TYPE(s, value);
        } else {
            throw "value has wrong format";
        }
        break;
    case "TupleStructType":
        if (is_TUPLE_STRUCT_TYPE(value)) {
            serialize_TUPLE_STRUCT_TYPE(s, value);
        } else {
            throw "value has wrong format";
        }
        break;
    case "EnumType":
        if (is_ENUM_TYPE(value)) {
            serialize_ENUM_TYPE(s, value);
        } else {
            throw "value has wrong format";
        }
        break;
    default:
        throw "type not implemented";
    }
    return s.finish();
}

exports.serialize = serialize
